lib/core/src/display/navigation/navigator.rs

```diff
@@ -0,0 +1,158 @@
+use crate::display::rendering::CameraType;
+use crate::display::rendering::DOMContainer;
+
+use nalgebra::Vector3;
+
+// =================
+// === Animation ===
+// =================
+
+/// Struct used to store animation related properties.
+struct Animation {
+    desired_pos  : Vector3<f32>,
+    vel          : Vector3<f32>,
+    desired_rate : f32,
+    time_count   : f32,
+    drag         : f32,
+    spring_coeff : f32,
+    mass         : f32,
+    min_vel      : f32
+}
+
+// =================
+// === Navigator ===
+// =================
+
+/// Navigator enables camera navigation with mouse interactions on the specified
+/// DOM.
+pub struct Navigator {
+    dom    : DOMContainer,
+    events : Rc<EventHandler>,
+    anim   : Animation
+}
+
+impl Navigator {
+    pub fn new(dom:&DOMContainer, camera:&Camera) -> Result<Self> {
+        let events        = EventHandler::new(dom)?;
+        let dom           = dom.clone();
+
+        let desired_pos   = *camera.position();
+        let vel           = Vector3::new(0.0, 0.0, 0.0);
+        let time_count    = 0.0;
+        let desired_rate  = 120.0; // 120fps
+        let drag          = 10.0;
+        let spring_coeff  = 1.5;
+        let mass          = 20.0;
+        let min_vel       = 0.001;
+
+        let animation = Animation {
+            desired_pos,
+            vel,
+            time_count,
+            desired_rate,
+            drag,
+            spring_coeff,
+            mass,
+            min_vel
+        };
+
+        Ok(Self { dom, events, anim: animation })
+    }
+
+    /// Polls for mouse events and navigates the camera.
+    pub fn navigate(&mut self, camera:&mut Camera, delta_seconds:f32) {
+        match self.events.poll() {
+            Event::Start      => self.anim.desired_pos = *camera.position(),
+            Event::Zoom(zoom) => self.zoom(camera, zoom),
+            Event::Pan(pan)   => self.pan(camera, pan),
+            Event::None       => ()
+        }
+        self.animate(camera, delta_seconds);
+    }
+
+    /// For zooming into a desired focused position.
+    fn zoom(&mut self, camera:&mut Camera, zoom:ZoomEvent) {
+        if let CameraType::Perspective(persp) = camera.camera_type() {
+            let point        = zoom.focus - self.dom.position();
+            // 0..1 normalization
+            let normalized_x = point.x / self.dom.dimensions().x;
+            let normalized_y = point.y / self.dom.dimensions().y;
+            // -1..1 normalization
+            let normalized_x = (normalized_x - 0.5) * 2.0;
+            let normalized_y = (normalized_y - 0.5) * 2.0;
+
+            let z =  zoom.amount * 6.0;
+            let x = -normalized_x * z / persp.aspect;
+            let y =  normalized_y * z / camera.get_y_scale();
+
+            let animation = &mut self.anim;
+            animation.desired_pos += Vector3::new(x, y, z);
+            if  animation.desired_pos.z < persp.near + 1.0 {
+                animation.desired_pos.z = persp.near + 1.0;
```

Comment:
This code contains a lot of hardcoded variables and operations which are not explained. What are for example these lines responsible for?

```rust
            let z =  zoom.amount * 6.0;
            let x = -normalized_x * z / persp.aspect;
            let y =  normalized_y * z / camera.get_y_scale();
```

Every part of the code which is not **imediatelly obvious** after reading, should be refactored to separate function. As @farmaazon always tells - code should be read as a book.

Reply:
@wdanilo this is done for compensating horizontal scaling for aspect adjustments and vertical scaling for fov.

```rust
            // Scale X and Y to compensate aspect and fov.
            let z =  zoom.amount * self.zoom_speed;
            let x = -normalized.x * z / persp.aspect;
            let y =  normalized.y * z / camera.get_y_scale();
```